e2e/nomostest/nt.go

```diff
@@ -755,70 +755,6 @@ func (nt *NT) portForwardGitServer() {
 	)
 }
 
-// portForwardRegistryServer forwards the registry-server deployment to a port.
-func (nt *NT) portForwardRegistryServer() {
```

Comment:
I feel I'd prefer to keep this close in proximity to the similar methods so that it's easier to discover. Is there a reason this code needs to be moved, or can we call `portForwardRegistryServer` from `setupRegistryClient`?

Reply:
Login needs to happen after Ready and before Push, and portForwardRegistryServer sets up Pushes from the callbacks that run when the registry Pod is Ready. 

Today this isn't critical because the local registry Login is a no-op, but if we ever add auth for the local registry, the login needs to happen after each time the registry is restarted, after it's ready and after the port-forwarding is ready.

Reply:
What I'm suggesting is just a refactoring, e.g.:
```golang
func setupRegistryClient(nt *NT, opts *ntopts.New) error {
    ...
    nt.portForwardRegistryServer(opts)
    ...
}
```

Reply:
Login needs to be performed at different times depending on whether there's port-forwarding or not.

Reply:
I understand that, here's the complete snippet I had in mind to clear ambiguity:
```golang
// portForwardRegistryServer forwards the registry-server deployment to a port.
func (nt *NT) portForwardRegistryServer(opts *ntopts.New) {
	// The local registry uses ephemeral storage. So re-push all the oci images
	// and helm charts after the registry recovers from a crash.
	var onReadyCallbacks []func(int, string)
	if opts.RequireOCIProvider && *e2e.OCIProvider == e2e.Local {
		provider := nt.OCIProvider.(*registryproviders.LocalOCIProvider)
		resetImages := func(newPort int, podName string) {
			for _, image := range nt.ociImages {
                if err := provider.Login(); err != nil {
			        nt.T.Fatalf("login: %v", err)
			    }
				// construct push URL with provided port. Calling LocalPort would lead to deadlock
				if err := image.Push(provider.PushURLWithPort(newPort, image.Name)); err != nil {
					nt.T.Fatalf("pushing image: %w", err)
				}
			}
		}
		onReadyCallbacks = append(onReadyCallbacks, resetImages)
	}
	if opts.RequireHelmProvider && *e2e.HelmProvider == e2e.Local {
		provider := nt.HelmProvider.(*registryproviders.LocalHelmProvider)
		resetPackages := func(newPort int, podName string) {
            if err := provider.Login(); err != nil {
			    nt.T.Fatalf("login: %v", err)
			}
			for _, pkg := range nt.helmPackages {
				// construct push URL with provided port. Calling LocalPort would lead to deadlock
				if err := pkg.Push(provider.PushURLWithPort(newPort, pkg.Name)); err != nil {
					nt.T.Fatalf("pushing image: %w", err)
				}
			}
		}
		onReadyCallbacks = append(onReadyCallbacks, resetPackages)
	}
	portForwarder := nt.newPortForwarder(
		TestRegistryNamespace,
		TestRegistryServer,
		fmt.Sprintf(":%d", RegistryHTTPPort),
		portforwarder.WithOnReadyCallback(func(newPort int, podName string) {
			for _, fn := range onReadyCallbacks {
				fn(newPort, podName)
			}
		}),
	)
	if opts.RequireOCIProvider && *e2e.OCIProvider == e2e.Local {
		provider := nt.OCIProvider.(*registryproviders.LocalOCIProvider)
		nt.T.Cleanup(func() {
			// clear PortForwarder after each test. at this point it has been stopped
			// a new PortForwarder is created for each test.
			provider.PortForwarder = nil
		})
		provider.PortForwarder = portForwarder
	}
	if opts.RequireHelmProvider && *e2e.HelmProvider == e2e.Local {
		provider := nt.HelmProvider.(*registryproviders.LocalHelmProvider)
		nt.T.Cleanup(func() {
			// clear PortForwarder after each test. at this point it has been stopped
			// a new PortForwarder is created for each test.
			provider.PortForwarder = nil
		})
		provider.PortForwarder = portForwarder
	}
	nt.startPortForwarder(
		TestRegistryNamespace,
		TestRegistryServer,
		portForwarder,
	)
}

...

func setupRegistryClient(nt *NT, opts *ntopts.New) error {
    if opts.RequireOCIProvider && *e2e.OCIProvider == e2e.Local || opts.RequireHelmProvider && *e2e.HelmProvider == e2e.Local {
        nt.portForwardRegistryServer()
    }
	// For non-local registries, just login before each test that uses the provider.
	if opts.RequireOCIProvider && *e2e.OCIProvider != e2e.Local {
		if err := nt.OCIProvider.Login(); err != nil {
			return err
		}
	}
	if opts.RequireHelmProvider && *e2e.HelmProvider != e2e.Local {
		if err := nt.HelmProvider.Login(); err != nil {
			return err
		}
	}

	// Register this cleanup after startPortForwarder to ensure this cleanup is
	// executed before the port-forwarder is shut down.
	nt.T.Cleanup(func() {
		if opts.RequireHelmProvider {
			// Delete all charts & images pushed during the test.
			for _, pkg := range nt.helmPackages {
				if err := pkg.Delete(); err != nil {
					nt.T.Error(err)
				}
			}
			nt.helmPackages = nil
			// Logout to reset the local client credentials.
			if err := nt.HelmProvider.Logout(); err != nil {
				nt.T.Error(err)
			}
		}
		if opts.RequireOCIProvider {
			// Delete all charts & images pushed during the test.
			for _, image := range nt.ociImages {
				if err := image.Delete(); err != nil {
					nt.T.Error(err)
				}
			}
			nt.ociImages = nil
			// Logout to reset the local client credentials.
			if err := nt.OCIProvider.Logout(); err != nil {
				nt.T.Error(err)
			}
		}
	})

	return nil
}
```

Reply:
Fine.

I don't like all the spaghetti caused by these callbacks, or that we need to put login in two different places, but cleaning that up seems like it would take a lot more refactoring and I'm not sure how to do it yet.

Reply:
This also means we need to pass the test options into an NT method, which feels kind of wrong. Usually for the NT methods the options have already been resolved into the new NT. But I don't really want to copy the opts variables into NT variables...

Reply:
Alright. I think I finally found a way to refactor this.

I added a new `ProxiedRegistryProvider` interface that has ProxyAddress, Restore, and Reset methods. This way the details of what's happening is hidden in the proxy-specific providers without causing deadlocks. As a bonus, the Cleanup is now separate for the proxy providers and non-proxy providers. So the Restore and Reset methods are both called from `portForwardRegistryServer` and `setupRegistryClient` only needs to call Login/Logout for the non-proxied providers.